### PR TITLE
WIP: UNI-306 J. Use JSON Schemas to create objects, and to verify object payloads. etc.

### PR DESCRIPTION
**Pull request by brev**
WIP!

UNI-306 J. Use JSON Schemas to create objects, and to verify object payloads. etc.

**Low priority, no rush**

- Use JSON Schemas to create initial object payloads
- Use JSON Schemas to validate payloads
- Create a new Model.json Schema for Model objects
- Move `schemas/index.js` to `database/schema/index.js` along with all of
  it's friends, and fix references.
- Moved `js/samples/` to just `samples/` since they belong to both PY + JS now.
- Unsupported test data moved to `tests/data/unsupported/` @marionleborgne.
- Fix a few final copyright dates

Lint, Docs, Tests, Run... all passing.

Please review @lscheinkman @marionleborgne @vitaly-krugl @cbaranski

### CODE REVIEW

File: unicorn/js/database/schema/Metric.json [hunk 1]
```diff
@@ -22,7 +22,7 @@
       "type": "string"
     },
     "format": {
-      "type": "string",
+      "type": ["null", "string"],
```
Comment (lscheinkman) [inline unicorn/js/database/schema/Metric.json]: Is it really necessary to add  `null` type to `optional` fields?
There are other schemas with optional fields

**Comment by lscheinkman** [comment]
@brev Could you also add tests covering the `schema instantiator` and `ModelStore`.
See `DatabaseService_test.js` for schema and `FileStore_test.js` for store examples.

**Comment by lscheinkman** [comment]
:+1: Great find on the `jsonschema` version difference between `py` and `js`

**Comment by brev** [comment]
@lscheinkman Working on the tests now, I'll update this later when ready for you to review again thanks

File: unicorn/js/database/schema/MetricData.json [hunk 2]
```diff
@@ -6,16 +6,18 @@
...
       "type": "string",
       "maxLength": 50,
-      "minLength": 1
+      "minLength": 1,
+      "required": true
     },
     "timestamp": {
       "type": "string",
-      "format": "date-time"
+      "format": "date-time",
```
Comment (lscheinkman) [inline unicorn/js/database/schema/MetricData.json]: JSON schema `date-time` validation conflicts with unicorn `date-time` validation and this schema will fail to validate when using date fields other than internet timestamps (http://tools.ietf.org/html/rfc3339). We should just remove the `format` property and let unicorn validate the `date-time` fields

File: unicorn/js/database/schema/ModelData.json [hunk 3]
```diff
@@ -6,19 +6,22 @@
...
       "type": "string",
       "maxLength": 50,
-      "minLength": 1
+      "minLength": 1,
+      "required": true
     },
     "timestamp": {
       "type": "string",
-      "format": "date-time"
+      "format": "date-time",
```
Comment (lscheinkman) [inline unicorn/js/database/schema/ModelData.json]: See https://github.com/numenta/numenta-apps/pull/671/files#r54631168

**Comment by lscheinkman** [comment]
:+1: LGTM, the `ModelStore` test can be updated in a later PR

File: unicorn/js/database/schema/MetricData.json (hunk 2 above)
Comment (brev) [inline unicorn/js/database/schema/MetricData.json]: thanks @lscheinkman I'll make a JIRA for this to make sure we come back to it soon

**Comment by brev** [comment]
@lscheinkman removed the `ModelStore` test from here, merging.

**Pull request merged**